src/chain/shared_state.rs

```diff
@@ -292,3 +321,66 @@ impl Debug for SectionProofChain {
         )
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{chain::SectionInfo, BlsPublicKey, FullId, Prefix, XorName};
+    use std::collections::BTreeSet;
+    use unwrap::unwrap;
+
+    fn gen_section_info(pfx: Prefix<XorName>) -> SectionInfo {
+        let sec_size = 5;
+        let mut members = BTreeSet::new();
+        for _ in 0..sec_size {
+            let id = FullId::within_range(&pfx.range_inclusive());
+            let _ = members.insert(*id.public_id());
+        }
+        unwrap!(SectionInfo::new(members, pfx, None))
+    }
+
+    fn gen_pk(pfx: Prefix<XorName>) -> BlsPublicKey {
+        BlsPublicKey::from_section_info(&gen_section_info(pfx))
+    }
+
+    #[test]
+    fn test_update_their_keys() {
+        let empty: Prefix<XorName> = Default::default();
+        let pfx_0 = empty.pushed(false);
+        let pfx_1 = empty.pushed(true);
+        let pfx_10 = pfx_1.pushed(false);
+        let pfx_11 = pfx_1.pushed(true);
+        let pfx_100 = pfx_10.pushed(false);
+        let pfx_101 = pfx_10.pushed(true);
+
+        // our section is 0
+        let sec_0 = gen_section_info(pfx_0);
+        let mut state = SharedState::new(sec_0);
+
+        // add keys for neighbours 10 and 11
+        let pk_sec_10 = gen_pk(pfx_10);
```

Comment:
What happen if we do not have pfx_10 or pfx_11?
What happen if we only have our side of the prefix when starting with pfx_0 (and we had a key when we were empty prefix).

Reply:
What do you mean by "our side of the prefix"?
Also note that this starts with not having 10 or 11 in `their_keys` - they are simply inserted, then.

Reply:
I meant something like.

Start empty prefix.
Split -> Prefix0 & Prefix1
We are Prefix0, we do not have key for Prefix1?

(not sure it is possible, I guess we would have both key at the time of the split and so we would also always add Prefix1?).

Reply:
I think the plan was that we'll have both keys from the DKGs we were running when preparing for the split (we'll have the private key share only from one of them, but we'll be able to obtain `PublicKeySet`s from both).

Reply:
So I guess the initial state should be:
```
        // our section is 0, so when we splited we have already added Prefix 1. 
        let sec_0 = gen_section_info(pfx_0);
        let mut state = SharedState::new(sec_0);
        let pk_sec_1 = gen_pk(pfx_1);
        state.update_their_keys(pfx_1, pk_sec_1.clone());
```

Reply:
Could be. But I think the one in the test is also valid - we might have just become an elder of 0, and 10 and 11 might have already existed. We would then initialise the chain in a manner similar to what is in the test.

Reply:
i'm not 100% sure that is the case, that bring a new thing to be careful about:
Are we doing the right thing when a new node join? (i.e is all the information necessary in Parsec).
Or after a split when I guess we are re-voting for all these sections info so new joining nodes get the right thing. Do we have the section re-consensused in order?

i.e should
```
             // Update their_keys in chain
            self.chain.update_their_keys(
                *sec_info.prefix(),
                BlsPublicKey::from_section_info(&sec_info),
            ); 
```

Take care of version number? Of parents consensused after children prefix?

Reply:
>Are we doing the right thing when a new node join? (i.e is all the information necessary in Parsec).

I'm not sure if it is there, but we certainly were planning for it to be there - the whole shared state is supposed to be included in the Genesis event eventually (maybe it already is, I'm not sure).

So a joining node should receive the initial shared state from the Genesis event, and then any updates would result from other consensused events, which it would receive as well.